View.elm

```diff
@@ -0,0 +1,304 @@
+module View exposing (..)
+
+import Html exposing (Html)
+import Html.Events
+
+
+-- # Elements
+-- This is a type level API for Html documents.
+-- Use the standard types defined here or create your own.
+
+
+type H1 child
+    = H1Type Never
+
+
+type H2 child
+    = H2Type Never
+
+
+type Section child
+    = SectionType Never
+
+
+type P child
+    = PType Never
+
+
+type OnClick msg child
+    = OnClickType Never
+
+
+type OneOf child1 child2
+    = OneOfType Never
+
+
+
+-- # Single type representation
+-- This is meant for creating view functions.
+-- 1. A phantom type ensures the View corresponds to the page type.
+-- 2. Customize the `SubView` type with your own stuff.
+-- 3. We don't render directly into Html to support different interpreters:
+--    - Elm-css uses a different Html type.
+--    - A view test might wish to interpret the view in a different way.
+--    - Certain optimizations (like collapsing unnecessary containers) are only
+--      possible if we can introspect our page structure.
+
+
+type View tipe custom msg
+    = View (SubView custom msg)
+
+
+type SubView custom msg
+    = H1 (SubView custom msg)
+    | H2 (SubView custom msg)
+    | Section (SubView custom msg)
+    | P (SubView custom msg)
+    | Text String
+    | List (List (SubView custom msg))
+    | OnClick msg (SubView custom msg)
+    | Custom custom (SubView custom msg)
+
+
+toSubView : View tipe custom msg -> SubView custom msg
+toSubView (View subView) =
+    subView
+
+
+h1 : View tipe custom msg -> View (H1 tipe) custom msg
+h1 child =
+    View <| H1 (toSubView child)
+
+
+h2 : View tipe custom msg -> View (H2 tipe) custom msg
+h2 child =
+    View <| H2 (toSubView child)
+
+
+section : View tipe custom msg -> View (Section tipe) custom msg
+section child =
+    View <| H2 (toSubView child)
+
+
+p : View tipe custom msg -> View (P tipe) custom msg
+p child =
+    View <| P (toSubView child)
+
+
+onClick : msg -> View tipe custom msg -> View (OnClick msg tipe) custom msg
```

Comment:
💭 maybe `on` like in `Html.Events`

Reply:
I added this, keeping `onClick`. The benefit of having `on` is that all events are supported. The benefit of using the `OnClick` type is that you have more type safety: we can check that it's really click events your implementing and not something else. Now you have the option.